2022Swerve/src/main/java/frc/robot/subsystems/AutonSubsytem.java

```diff
@@ -0,0 +1,137 @@
+
+    private AutonModes autonMode;
+
+    public AutonSubsytem(SwerveSubsystem swerveSubsystem){
+        this.swerveSubsystem = swerveSubsystem;
+
+        //Add auton modes to chooser
+        for(AutonModes mode : AutonModes.values()){
+            modeChooser.addOption(mode.toString(), mode);
+        }
+        configTab.add("Auton mode", modeChooser);
+
+        //Define PID controllers for tracking trajectory
+        xController = new PIDController(AutonConstants.kXTranslationP, 0, 0);
+        yController = new PIDController(AutonConstants.kYTranslationP, 0, 0);
+        rotationController = new ProfiledPIDController(
+            AutonConstants.kRotationP, 0, 0, AutonConstants.kRotationalConstraints);
+        rotationController.enableContinuousInput(-Math.PI, Math.PI);
+    }
+
+    private Trajectory getTrajectory(String name){
+        Trajectory trajectory;
+        try {
+            Path trajectoryPath = Filesystem.getDeployDirectory().toPath().resolve("pathplanner/generatedJSON/" + name + ".wplib.json");
+            trajectory = TrajectoryUtil.fromPathweaverJson(trajectoryPath);
+        } catch (IOException ex) {
+            DriverStation.reportError("Unable to open trajectory: " + name, ex.getStackTrace());
+            trajectory = null;
+        }
+        return trajectory;
+    }
+
+    private SwerveControllerCommand followTrajectory(String name){
+        trajectoryLog.append(name);
+        return new SwerveControllerCommand(
+            getTrajectory(name),
+            swerveSubsystem::getPose, 
+            DriveConstants.kDriveKinematics, 
+            xController, 
+            yController, 
+            rotationController, 
+            swerveSubsystem::setModuleStates, 
+            swerveSubsystem
+        );
+    }
+
+    private Command resetOdometry(String initialTrajectory){
+        //Resets odometry to the initial position of the given trajectory
+        return new InstantCommand(() -> swerveSubsystem.resetOdometry(getTrajectory(initialTrajectory).getInitialPose()));
+    }
+    
+    private Command getAutonSequence(){
+        autonMode = modeChooser.getSelected();
+        //Sequence of actions to be performed during the autonomous period
+        switch(autonMode){
+            case FORWARD:
+               return 
+                    new SequentialCommandGroup(
+                        resetOdometry("Forward"),
+                        followTrajectory("Forward")
+                    );
+            case BACKWARD:
+                return
+                    new SequentialCommandGroup(
+                        resetOdometry("Backward"),
+                        new WaitCommand(3),
+                        followTrajectory("Backward")
+                    );
+            case FORWARD180:
+                return
+                    new SequentialCommandGroup(
+                        resetOdometry("Forward180"),
+                        followTrajectory("Forward180")
```

Comment:
What do you think about an align method?  Something that uses the current pose and inputs a rotation and does it?

That would make this forward, then align(180).

It would be good because you can then expand it to alignToTarget() Or something that determines rotation to a target...
And would be a good example to have in the code

Reply:
> That would make this forward, then align(180).

It would stay a trajectory, as the trajectory does the rotation while translating (like a swerve drive), instead of translating then rotating (like a tank drive).

I think an align method should be something we use when we have a sensor to use, but as it stands, the autonomous planner makes it very easy to align the robot towards a specific direction.

Reply:
Ya, but by the time you have a sensor, it would be nice to have the component pieces built out.

I am envisioning something like this psuedocode:
align degrees:
  read current pose
  get current rotation
  add degrees to rotation
  Mark that we are off of our ending pose.

( Because the next trajectory will have to translate the starting pose to this slightly different one. )

Reply:
and then when you get a sensor, it can be:
alignToTarget:
  Use vision to find needed rotation from where we are
  align targetDegrees

Reply:
Maybe this would also be best as a follow on feature?

Reply:
I will create a sensor alignment method, sure. It just wouldn't be "turn an amount of degrees", it would align to a heading/value on the sensor (gyro in this case).

Reply:
https://trello.com/c/NcE4SjuI

my thought is that keeping them separate would allow for better encapsulation